File: scripts/refine-tools.mjs
#!/usr/bin/env node
/**
 * Refine 3 unrefined tools in app/data/tools.ts using Qwen API.
 * Reads the file, identifies tools with short longDescriptions, refines them.
 */
import { readFileSync, writeFileSync } from 'fs';
import { fileURLToPath } from 'url';
import path from 'path';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const TOOLS_FILE = path.join(__dirname, '..', 'app', 'data', 'tools.ts');

// Load env
import { createRequire } from 'module';
const require = createRequire(import.meta.url);

// We'll use the env variable
const QWEN_API_KEY = process.env.QWEN_API_KEY_1;

const QWEN_BASE_URL = 'https://dashscope.aliyuncs.com/compatible-mode/v1';
const MODEL = 'qwen-max';  // Use qwen-max for best quality

async function refineWithQwen(tool) {
  const systemPrompt = `You are a professional video tools reviewer and content writer for VidioPicks.com. 
Your job is to write a thorough, detailed, informative longDescription for a video-related tool.
The longDescription must be:
1. FACTUAL and DETAILED — include specific features, use cases, benchmarks where relevant
2. STRUCTURED — use multiple paragraphs with clear headings/explanations
3. PROFESSIONAL in tone but accessible
4. Between 300-600 words
5. Written in clear English (markdown supported)

Write the longDescription content only, no extra commentary.`;

  const userPrompt = `Write a professional, detailed longDescription for the following video tool:

Tool Name: ${tool.name}
Category: ${tool.category}
Current Description: ${tool.description}
Current Features: ${tool.features ? tool.features.join(', ') : 'N/A'}
Current Pricing: ${tool.pricing} - ${tool.pricingDetail}
Use Case: ${tool.useCase}

The longDescription should be comprehensive (300-600 words), markdown-formatted, multi-paragraph, covering:
- What the tool is and who it's for
- Key features and capabilities with specific details
- Performance or quality benchmarks if applicable
- Platform support and integration
- Ideal use cases

Write ONLY the longDescription content. Use backtick-delimited string format suitable for a TypeScript template literal.`;

  const response = await fetch(`${QWEN_BASE_URL}/chat/completions`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${QWEN_API_KEY}`
    },
    body: JSON.stringify({
      model: MODEL,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userPrompt }
      ],
      temperature: 0.7,
      max_tokens: 2048
    })
  });

  if (!response.ok) {
    const errText = await response.text();
    throw new Error(`Qwen API error: ${response.status} ${errText}`);
  }

  const data = await response.json();
  return data.choices[0].message.content.trim();
}

async function main() {
  // Tools to refine (by id) — picked the shortest longDescriptions
  const targetTools = [
    'final-cut-pro',
    'davinci-resolve',
    'capcut'
  ];

  const content = readFileSync(TOOLS_FILE, 'utf-8');
  let updatedContent = content;

  for (const targetId of targetTools) {
    // Find the tool in the file
    const idRegex = new RegExp(`id:\\s*"${targetId}"`);
    const idMatch = idRegex.exec(updatedContent);
    if (!idMatch) {
      console.error(`Tool "${targetId}" not found in file`);
      continue;
    }

    // Find the start position of this tool object
    const toolStartIdx = updatedContent.lastIndexOf('{', idMatch.index);
    if (toolStartIdx < 0) {
      console.error(`Could not find object start for "${targetId}"`);
      continue;
    }

    // Extract the tool info for API call
    const toolBlock = updatedContent.substring(toolStartIdx, toolStartIdx + 4000);
    
    // Parse tool data roughly
    const nameMatch = toolBlock.match(/name:\s*"([^"]+)"/);
    const catMatch = toolBlock.match(/category:\s*"([^"]+)"/);
    const descMatch = toolBlock.match(/description:\s*"([^"]+)"/);
    const featuresMatch = toolBlock.match(/features:\s*\[([^\]]+)\]/);
    const pricingMatch = toolBlock.match(/pricing:\s*"([^"]+)"/);
    const pricingDetailMatch = toolBlock.match(/pricingDetail:\s*"([^"]+)"/);
    const useCaseMatch = toolBlock.match(/useCase:\s*"([^"]+)"/);

    const tool = {
      id: targetId,
      name: nameMatch?.[1] || targetId,
      category: catMatch?.[1] || '',
      description: descMatch?.[1] || '',
      features: featuresMatch ? featuresMatch[1].split(',').map(f => f.trim().replace(/"/g, '')) : [],
      pricing: pricingMatch?.[1] || '',
      pricingDetail: pricingDetailMatch?.[1] || '',
      useCase: useCaseMatch?.[1] || ''
    };

    console.log(`\n=== Refining: ${tool.name} (${tool.id}) ===`);
    
    // Find the existing longDescription
    const longDescRegex = new RegExp(`id:\\s*"${targetId}"[\\s\\S]*?longDescription:\\s*\`([\\s\\S]*?)\``);
    const longDescMatch = longDescRegex.exec(updatedContent);
    
    if (!longDescMatch) {
      console.error(`Could not find longDescription for "${targetId}"`);
      continue;
    }

    const newLongDesc = await refineWithQwen(tool);
    console.log(`Generated ${newLongDesc.length} chars of new content`);

    // Replace the longDescription
    const oldLongDesc = longDescMatch[1];
    // Escape special regex chars in old string
    const escapedOld = oldLongDesc.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    
    // Find the exact full longDescription (with backticks)
    const fullMatch = updatedContent.match(new RegExp(`(longDescription:\\s*\`)[\\s\\S]*?(\`\\s*,)`));
    
    if (!fullMatch) {
      console.error(`Could not parse longDescription boundaries for "${targetId}"`);
      continue;
    }

    // Debug: find the exact position
    const longDescPrefix = `longDescription: \``;
    const prefixIdx = updatedContent.indexOf(longDescPrefix, toolStartIdx);
    if (prefixIdx < 0) {
      console.error(`Could not find longDescription: prefix for "${targetId}"`);
      continue;
    }

    const openTickIdx = prefixIdx + longDescPrefix.length - 1; // position of opening backtick
    const closeTickIdx = updatedContent.indexOf('`', openTickIdx + 1);
    if (closeTickIdx < 0) {
      console.error(`Could not find closing backtick for "${targetId}"`);
      continue;
    }

    const before = updatedContent.substring(0, openTickIdx);
    const after = updatedContent.substring(closeTickIdx + 1);

    updatedContent = before + newLongDesc + after;
    console.log(`✅ Refined ${tool.name} successfully`);
  }

  writeFileSync(TOOLS_FILE, updatedContent, 'utf-8');
  console.log('\n✅ All refinements saved to tools.ts');
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: refine_tools.py
#!/usr/bin/env python3
"""Refine short tool entries via Qwen API using field-by-field replacement."""
import json
import re
import sys
import os

sys.path.insert(0, '/home/edi/vidiopicks-hub')

# Read API key
with open('/home/edi/.hermes/api_keys.env', 'rb') as f:
    raw = f.read()
idx = raw.find(b'QWEN_API_KEY_1=')
start = idx + len(b'QWEN_API_KEY_1=')
end = raw.find(b'\n', start)
QWEN_API_KEY = raw[start:end].decode()

# Read the tools.ts file
with open('/home/edi/vidiopicks-hub/app/data/tools.ts', 'r', encoding='utf-8') as f:
    tools_content = f.read()

def call_qwen(prompt, system="You are a professional video production tool reviewer. Generate detailed, authoritative G2-style content."):
    """Call Qwen API and return response text."""
    import urllib.request
    import urllib.error
    
    data = json.dumps({
        "model": "qwen-plus",
        "messages": [
            {"role": "system", "content": system},
            {"role": "user", "content": prompt}
        ],
        "temperature": 0.7,
        "max_tokens": 3000
    }).encode('utf-8')
    
    req = urllib.request.Request(
        "https://dashscope.aliyuncs.com/compatible-mode/v1/chat/completions",
        data=data,
        headers={
            "Authorization": f"Bearer {QWEN_API_KEY}",
            "Content-Type": "application/json"
        }
    )
    
    try:
        import subprocess
        result = subprocess.run(
            ['curl', '-s', '--proxy', 'http://127.0.0.1:7890',
             'https://dashscope.aliyuncs.com/compatible-mode/v1/chat/completions',
             '-H', f'Authorization: Bearer {QWEN_API_KEY}',
             '-H', 'Content-Type: application/json',
             '-d', data.decode('utf-8')],
            capture_output=True, text=True, timeout=120
        )
        resp = json.loads(result.stdout)
        return resp['choices'][0]['message']['content']
    except Exception as e:
        print(f"API error: {e}")
        return None


def get_existing_entry(content, tool_id):
    """Extract the full tool entry from content by matching the id."""
    # Find the tool entry by id
    pattern = r'(\s*\{\s*\n\s*id:\s*"' + re.escape(tool_id) + r'")'
    match = re.search(pattern, content)
    if not match:
        print(f"Could not find tool id: {tool_id}")
        return None
    start_pos = match.start()
    
    # Find the closing bracket at matching depth
    depth = 0
    in_string = False
    string_char = None
    in_template = False
    pos = start_pos
    while pos < len(content):
        ch = content[pos]
        if in_template:
            if ch == '`':
                in_template = False
        elif in_string:
            if ch == '\\':
                pos += 2
                continue
            if ch == string_char:
                in_string = False
        else:
            if ch == '`':
                in_template = True
            elif ch in '"\'':
                in_string = True
                string_char = ch
            elif ch == '{':
                depth += 1
            elif ch == '}':
                depth -= 1
                if depth == 0:
                    # Found the closing bracket
                    return content[start_pos:pos+1], start_pos, pos+1
        pos += 1
    
    print(f"Could not find closing brace for tool: {tool_id}")
    return None


def generate_new_entry(tool_id, tool_name, category, rating, review_count, current_props):
    """Generate new refined content for a tool via Qwen."""
    
    prompt = f"""Generate a complete, detailed G2-style tool entry for "{tool_name}" (id: {tool_id}, category: {category}).

Current entry (short/needs expansion). Generate ALL new fields below with high-quality content.

Generate the following fields in raw JSON format (EXACT structure):
1. longDescription: A 1200-1500 character markdown description with:
   - Overview paragraph (who it's for, what it does)
   - ### Key Features section with 4-5 bullet-pointed feature descriptions
   - A benchmark/performance paragraph with specific numbers
   - Platform support details

2. pros: Array of 5-7 specific, actionable pros (each 5-15 words)

3. cons: Array of 2-4 specific cons

4. pricing: Short label like "Subscription", "Freemium", "Paid", "Free", "Open Source"

5. pricingDetail: Detailed pricing info mentioning plans and prices

6. features: Array of 6-10 specific features

7. scoreBreakdown: Object with features, reviews, momentum, popularity (all numbers 70-98)

8. userQuotes: Array of 3 objects with role, company, quote

Current info:
- Rating: {rating}
- Review count: {review_count}

Return ONLY a JSON object with these fields. No markdown, no code fences."""
    
    result = call_qwen(prompt)
    if not result:
        return None
    
    # Parse JSON from response
    # Try to find JSON in the response
    json_match = re.search(r'\{[\s\S]*\}', result)
    if json_match:
        try:
            data = json.loads(json_match.group())
            return data
        except json.JSONDecodeError as e:
            print(f"JSON parse error for {tool_id}: {e}")
            print(f"Raw response snippet: {result[:200]}")
            return None
    else:
        print(f"No JSON found in response for {tool_id}")
        return None


def format_new_entry_string(tool_id, tool_name, category, rating, review_count, icon_name, description, data, existing_entry_start):
    """Format a complete tool entry string, preserving the original structure."""
    
    # Extract the id, name, category etc. from the original entry start
    # We'll rebuild the entry keeping certain fields
    price = data.get('pricing', 'Paid')
    pricing_detail = data.get('pricingDetail', '')
    
    # Format longDescription - it uses backtick template literals
    long_desc = data.get('longDescription', '')
    
    pros = data.get('pros', [])
    cons = data.get('cons', [])
    features = data.get('features', [])
    score = data.get('scoreBreakdown', {})
    quotes = data.get('userQuotes', [])
    use_case = data.get('useCase', '')
    
    # Format pros array
    pros_str = json.dumps(pros, ensure_ascii=False)
    cons_str = json.dumps(cons, ensure_ascii=False)
    features_str = json.dumps(features, ensure_ascii=False)
    
    # Format score breakdown
    score_str = json.dumps(score, ensure_ascii=False)
    
    # Format quotes
    quotes_lines = []
    for q in quotes:
        role = q.get('role', 'User')
        company = q.get('company', 'Company')
        quote_text = q.get('quote', '')
        # Escape special chars
        quote_text = quote_text.replace('\\', '\\\\').replace('`', "'").replace('${', '\\${')
        quotes_lines.append(f'''      {{
        role: "{role}",
        company: "{company}",
        quote: "{quote_text}",
      }}''')
    quotes_str = '\n'.join(quotes_lines)
    
    # Build complete entry
    entry = f'''  {{
    id: "{tool_id}",
    name: "{tool_name}",
    category: "{category}",
    rating: {rating},
    reviewCount: {review_count},
    icon: {icon_name},
    description: "{description}",
    longDescription: `{long_desc}`,
    pros: {pros_str},
    cons: {cons_str},
    pricing: "{price}",
    pricingDetail: "{pricing_detail}",
    features: {features_str},
    useCase: "{use_case}",
    websiteUrl: "https://www.example.com/{tool_id}",
    alternatives: [
        "adobe-premiere-pro",
        "davinci-resolve"
      ],
    scoreBreakdown: {score_str},
    userQuotes: [
{quotes_str}
    ],
  }}'''
    
    return entry


# Tools to refine
tools_to_refine = [
    {
        "id": "adobe-after-effects",
        "name": "Adobe After Effects",
        "category": "Motion Graphics and Animation",
        "rating": 4.7,
        "reviewCount": 3240,
        "icon": "Video",
        "description": "Industry-standard motion graphics and visual effects software for film, TV, and web.",
    },
    {
        "id": "camtasia",
        "name": "Camtasia",
        "category": "Screen Recording",
        "rating": 4.5,
        "reviewCount": 3120,
        "icon": "Monitor",
        "description": "All-in-one screen recorder and video editor designed for professionals creating tutorials and presentations.",
    },
    {
        "id": "storyblocks",
        "name": "Storyblocks",
        "category": "Stock Video and Assets",
        "rating": 4.5,
        "reviewCount": 4120,
        "icon": "Video",
        "description": "All-in-one subscription platform for stock video, audio, images, and templates with AI-powered search.",
    },
]

for tool in tools_to_refine:
    print(f"\n=== Refining {tool['name']} ===")
    
    tid = tool['id']
    existing, start, end = get_existing_entry(tools_content, tid)
    if not existing:
        print(f"SKIP: Could not find entry for {tid}")
        continue
    
    print(f"Found entry at position {start}-{end}")
    print(f"Existing entry length: {len(existing)} chars")
    
    # Generate new content
    new_data = generate_new_entry(
        tid, tool['name'], tool['category'],
        tool['rating'], tool['reviewCount'],
        {}
    )
    
    if not new_data:
        print(f"SKIP: Failed to generate content for {tid}")
        continue
    
    # Build the new entry string
    new_entry = format_new_entry_string(
        tid, tool['name'], tool['category'],
        tool['rating'], tool['reviewCount'],
        tool['icon'], tool['description'], new_data,
        existing
    )
    
    # Verify brace balance
    open_braces = new_entry.count('{')
    close_braces = new_entry.count('}')
    print(f"Brace balance: open={open_braces}, close={close_braces}")
    
    if open_braces == close_braces:
        # Replace in content
        tools_content = tools_content[:start] + new_entry + tools_content[end:]
        print(f"✓ Successfully replaced entry for {tid}")
    else:
        print(f"✗ Brace mismatch for {tid}, skipping replacement")
        continue

# Final brace balance check
total_open = tools_content.count('{')
total_close = tools_content.count('}')
print(f"\nFinal brace balance: open={total_open}, close={total_close}")

if total_open == total_close:
    # Write the updated file
    with open('/home/edi/vidiopicks-hub/app/data/tools.ts', 'w', encoding='utf-8') as f:
        f.write(tools_content)
    print("✓ Written tools.ts successfully")
else:
    print("✗ Brace mismatch, NOT writing tools.ts")
